src/tfr_control/src/controller.cpp

```diff
@@ -30,20 +30,33 @@ namespace tfr_control
         registerInterface(&joint_effort_interface);
     }
 
-    void Controller::read() {
+    // Return true if bin is finished moving to its new position.
+    // TODO: A more explicit return type might be better here, or a possible redesign.
+    //       If other values read end up needing to be returned, then this should 
+    //       probably return some data structure. It may be that read values get published
+    //       out on topics instead.
+    bool Controller::read() 
```

Comment:
Let's make this method name a little more specific. If this is supposed to be the controller for all of the hardware interfaces on the whole robot, then methods just like `read` or `write` are not descriptive enough. Maybe `read_bin` or `write_bin` might improve maintainability.

Reply:
Follow up question is `controller.cpp` the only controller class for the whole robot?

Reply:
`controller.cpp` is the class we use as an interface between the hardware and our message-passing system, though that interface is also shared by a ROS class called `controller_manager` (both of these classes, as well as my bin control server, live in the `controller_launcher` node).  So yes; `controller.cpp` is for the entire robot.  `read()` reads from our encoders (all of them) and does some as-yet-undefined action with the data.  `write()` writes voltage values out to the I2C connection for all our actuators. 

This was all set up by @AdamuKaapan, so he can correct me if I've gotten anything wrong.

Reply:
Yeah, that description looks correct. I think that we might want to find a different way to pass out the return value, whether we save them in some other local variable, put them out on topics, or something else. While I can't *think* of anything else that might need a return value from this method right now, it still feels like a kinda kludgey design and would prefer to find a better solution. And @collinsc due to all of the sensor inputs coming from the same hardware device it makes it much simpler/resource efficient to put it in one method, as much as I want to split it apart (although we could perform separate processing of said inputs on a separate input).

Reply:
Although now that I've seen the rest and seen that the bin control server is integrated with the controller_launcher itself, this isn't quiiiite as bad. I think I'd still prefer a slightly better solution though.